### PR TITLE
Fix: scheduler: do not enforce resource stop on a rejoined node that was the target of a failed migrate_to

### PR DESCRIPTION
**Pull request by gao-yan**
Previously given a scenario:

- rscA failed to migrate from node1 to node2 with a failed migrate_to on
node1
- rscA failed to stop on node2
- node2 got fenced and rscA recovered on node1
- node2 rejoined

, rscA would be considered being "multiple-active" and an unnecessary
full recovery would be issued.

With the commits, when node2 rejoins, rather then enforcing a likely
unnecessary stop of rscA on it, we wait for the probe of it to be
complete, then we will have a better idea about what we should do with it.

This is another attempt to address the issue originally brought up from:
https://github.com/ClusterLabs/pacemaker/pull/2146#pullrequestreview-911922137

### CODE REVIEW

File: lib/pengine/unpack.c [hunk 1]
```diff
@@ -2761,7 +2783,14 @@ unpack_migrate_to_failure(pe_resource_t *rsc, pe_node_t *node, xmlNode *xml_op,
                                       source, TRUE, data_set);
     target_migrate_from_id = pe__call_id(target_migrate_from);
 
-    if ((target_stop == NULL) || (target_stop_id < target_migrate_from_id)) {
+    if (/* If the resource has no lrm_rsc_op history on the target,
+         * that means its state is unknown there and it will be probed.
+         * Don't just consider it running there. We will get back here anyway in
+         * case the probe detects it's running there.
+         */
```
Comment (kgaillot) [inline lib/pengine/unpack.c:2797]: Unfortunately, we can't assume there will be a probe. There's resource-discovery (which we don't have easy access to at this point), or something could end up blocking the probe (maintenance mode on relevant node, relevant node is remote node whose connection is blocked, etc.). In that case, the resource could potentially be left multiply active.

However that's not too different from the situation without migration when a resource is already active on one node and a probe is blocked on some other node, so I suppose we can keep this. I can't think of anything better.

File: lib/pengine/unpack.c [hunk 2]
```diff
@@ -2790,6 +2797,11 @@ unpack_migrate_to_failure(pe_resource_t *rsc, pe_node_t *node, xmlNode *xml_op,
          */
         target_lrm_resource != NULL
         && first_named_child(target_lrm_resource, XML_LRM_TAG_RSC_OP) != NULL
+        /* If there's any probe/monitor operation on the target newer than this
+         * failed migrate_to indicating it was not running there, this migrate_to
+         * failure no longer matters for the target.
+         */
+        && ((target_monitor == NULL) || (is_newer_op(target_monitor, xml_op) <= 0))
```
Comment (kgaillot) [inline lib/pengine/unpack.c:2804]: I'm also a little concerned that we're introducing a tighter dependency on synchronized clocks, but again it seems reasonable and I don't have a better idea.

**Review by kgaillot — COMMENTED** [review]
Whew, this turned out to be quite the brain-twister. Good job figuring it out ...

I left some minor comments but I think it's good to go

File: lib/pengine/unpack.c (hunk 1 above)
Comment (gao-yan) [reply lib/pengine/unpack.c:2827]: Indeed. Functionized it now.

File: include/crm/pengine/internal.h [hunk 3]
```diff
@@ -431,6 +431,7 @@ extern void pe_free_action(pe_action_t * action);
 extern void resource_location(pe_resource_t * rsc, pe_node_t * node, int score, const char *tag,
                               pe_working_set_t * data_set);
 
+extern int is_newer_op(const xmlNode *xml_a, const xmlNode *xml_b);
```
[the remark this answers is not in the document]
Comment (gao-yan) [reply include/crm/pengine/internal.h]: Prefixed now.

File: lib/pengine/utils.c [hunk 4]
```diff
@@ -1744,7 +1744,11 @@ is_newer_op(const xmlNode *xml_a, const xmlNode *xml_b)
     const char *a_xml_id = crm_element_value(xml_a, XML_ATTR_ID);
     const char *b_xml_id = crm_element_value(xml_b, XML_ATTR_ID);
 
-    if (pcmk__str_eq(a_xml_id, b_xml_id, pcmk__str_casei)) {
+    const char *a_node = crm_element_value(xml_a, XML_LRM_ATTR_TARGET);
+    const char *b_node = crm_element_value(xml_b, XML_LRM_ATTR_TARGET);
+    bool same_node = pcmk__str_eq(a_node, b_node, pcmk__str_casei);
+
+    if (same_node && pcmk__str_eq(a_xml_id, b_xml_id, pcmk__str_casei)) {
```
[the remark this answers is not in the document]
Comment (gao-yan) [reply lib/pengine/utils.c]: Good catch.

File: lib/pengine/unpack.c (hunk 2 above)
Comment (gao-yan) [reply lib/pengine/unpack.c]: Indeed. Functionalized as well now.

File: lib/pengine/utils.c [hunk 5]
```diff
@@ -1766,13 +1767,14 @@ sort_op_by_callid(gconstpointer a, gconstpointer b)
...
+    } else if (same_node && a_call_id >= 0 && a_call_id < b_call_id) {
         sort_return(-1, "call id");
 
-    } else if (b_call_id >= 0 && a_call_id > b_call_id) {
+    } else if (same_node && b_call_id >= 0 && a_call_id > b_call_id) {
         sort_return(1, "call id");
 
-    } else if (b_call_id >= 0 && a_call_id == b_call_id) {
+    } else if (a_call_id >= 0 && b_call_id >= 0
+               && (!same_node || a_call_id == b_call_id)) {
```
Comment (gao-yan) [inline lib/pengine/utils.c:1777]: Revised the logic here to make it more consistent for the case of `!same_node`.

**Review by gao-yan — COMMENTED** [review]
Thanks for review and opinions.

**Pull request merged**